designs/odcr.md

```diff
@@ -0,0 +1,121 @@
+
+Pros: 
+- This puts the onus of checking on the user. It doesn't require checking if a Capacity Reservation satisfy the requirements of an instance type by assuming the users is doing this. It makes no assumption about what a Capacity Reservation can do nor implies can do. This is helpful because if there is expectation that Karpenter does this we may be asked to
+support checking if there is capacity in the reservation before launching nodes which would significantly increase the scope of this design.
+
+Cons:
+- This implementation is overly restrictive causing it to be difficult to use. Its possible that the restrictions makes it too hard for users to use EC2NodeClasses effectively.
+
+Both a pro and a con:
+- This forces users who wish to leverage fallback and prioritization to use NodePool weights rather than relying on EC2NodeClass. I can't tell if this is a good or a bad thing but in my
+  opinion it isn't always clear what Karpenter wants users to do in this aspect.
+
+#### Labels
+
+When a node is launched against a CapacityReservation we will expose Capacity Reservation
+information as labels `karpenter.k8s.aws/capacity-reservation-id` and `karpenter.k8s.aws/capacity-reservation-setting`.
+
+```yaml
+Name:               example-node
+Labels:             karpenter.k8s.aws/capacity-reservation-id=cr-12345
+                    karpenter.k8s.aws/capacity-reservation-setting=open
+                    karpenter.sh/capacity-type=on-demand
+```
+
+`karpenter.k8s.aws/capacity-reservation-id` will be the capacity reservation the node launched from. `karpenter.k8s.aws/capacity-reservation-setting` will depend on the launch template's `capacityReservationSpec`. It will either be a preference or a target.
+
+We will propagate this information via [instance](https://github.com/aws/karpenter-provider-aws/blob/main/pkg/providers/instance/types.go#L29) by extracting it from [DescribeInstance](https://github.com/aws/karpenter-provider-aws/blob/main/pkg/batcher/describeinstances.go#L48) [aws doc]([https://docs.aws.amazon.com/sdk-for-go/api/service/ec2/#EC2.DescribeInstances](https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_CapacityReservationSpecificationResponse.html)https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_CapacityReservationSpecificationResponse.html).
+
+### Failed to launch Nodes into Capacity Reservation
+
+The main failure scenario is when Capacity Reservation limit is hit and no new nodes can be launched from any Capacity Reservation the launch template targets. 
+
+#### Status
+
+We will expose this information both in the NodeClaim's, EC2NodeClass' and NodePool's Condition Status. _I know that currently NodeClasses and NodePool status do not have conditions but I wanted to see if we are opened to adding Conditions to these resources_
+
+```yaml
+Status:
+  Conditions:
+    Last Transition Time:  2024-02-22T03:46:42Z
+    Status:                LimitExceeded
```

Comment:
Do you think that we need this status condition if Karpenter is aware of the available capacity and it can just know when to fallback to other NodePools since it ran out of capacity?

Reply:
I think so. From an operator's perspective it would be good to know currently this nodeclass hit its limit and we are going to fallback into other node pools